### PR TITLE
[NOT FOR MERGE] VIP 3D audio implementation. Have known bugs.

### PR DESCRIPTION
**Pull request by pirogronian**
However, if someone really wants... it's for now mergable, but hard to debug, due to lack of Lua interface. Interface will go after merging with LuaWrapper2 branche, which will go with its own PR.

Mentioned bugs are strange behaviour of sound. Maybe due to my misunderstanding of OpenAl parameters.

Tests can be done by editing qtappwin.cpp as follows:
`index d84e7af9..331cd207 100644
--- a/src/celestia/qt/qtappwin.cpp
+++ b/src/celestia/qt/qtappwin.cpp
@@ -407,9 +407,35 @@ void CelestiaAppWindow::init(const QString& qConfigFileName,
     QTimer *t = new QTimer(dynamic_cast<QObject *>(this));
     QObject::connect(t, SIGNAL(timeout()), SLOT(celestia_tick()));
     t->start(0);
+    
+    
+    
+/*    t = new QTimer(this);
+    QObject::connect(t, &QTimer::timeout, this, &CelestiaAppWindow::dumpAudio);
+    t->start(1000);*/
+    QAction *a = new QAction("Debug", this);
+    a->setShortcut(QKeySequence("Ctrl+Alt+D"));
+    connect(a, &QAction::triggered, this, &CelestiaAppWindow::dumpAudio);
+    viewMenu->addAction(a);
+
+    Audio3D::Manager::setDopplerFactor(2);
+    m_appCore->audio3dManager.setGlobalDistanceScale(0.0001);
+    Selection sel = m_appCore->getSimulation()->findObjectFromPath("moon");
+    cout << sel.getName() << endl;
+    AudioObject *o = m_appCore->audio3dManager.getAudioObject(sel);
+    o->setTone(AudioTone::standardTone(432));
+    //o->setPitch(100);
+    //o->setGain(100);
+    o->setMaxDistance(1000000);
+    o->setLooping(true);
+    o->play();
 }
 
 
+void CelestiaAppWindow::dumpAudio() {
+    m_appCore->audio3dManager.dump(m_appCore->getSimulation()->getTime());
+}
+
 /*! Set up the application data directory, creating it if necessary. The
  *  directory contains user-specific, persistent information for Celestia
  *  (such as bookmarks) which aren't stored in settings. The location`

Then, run Celestia_QT, goto "Moon", should hear "bzzz" at 432 Hz, then slightly accelerate and try to rotate observer direction. Doppler effect appear to be twiced relatively to rotation circle. Is the same in opposite directions (180 deg) and opposite at orthogonal (90 deg).

### CODE REVIEW

File: src/audio3d/Audio3DBuffer.h [hunk 1]
```diff
@@ -0,0 +1,113 @@
+#ifndef __PIROGRONIAN__AUDIO_3D__BUFFER__H__
```
Comment (375gnu) [inline src/audio3d/Audio3DBuffer.h]: This looks like vanity.

File: src/audio3d/Audio3DConverts.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#ifndef __PIROGRONIAN__AUDIO_3D__CONVERTS__H__
+#define __PIROGRONIAN__AUDIO_3D__CONVERTS__H__
+
+#include <Eigen/Core>
+#include <AL/al.h>
+
+namespace Audio3D 
+{
+
+    Eigen::Vector3d inline convert(const ALfloat *v)
```
Comment (375gnu) [inline src/audio3d/Audio3DConverts.h]: Vector3f?

File: src/audio3d/Audio3DConverts.h [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        return Eigen::Vector3d(v[0], v[1], v[2]);
```
Comment (375gnu) [inline src/audio3d/Audio3DConverts.h]: no bounds checking?

Comment (375gnu) [reply src/audio3d/Audio3DConverts.h]: and `Eigen::Map<Eigen::Vecto3f>(v)` will be better

File: src/audio3d/Audio3DConverts.h (hunk 2 above)
Comment (375gnu) [reply src/audio3d/Audio3DConverts.h]: and you should use better name, convert is too generic

File: src/audio3d/Audio3DConverts.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    void inline convert (ALfloat *alv, const Eigen::Vector3d &ev)
+    {
+        alv[0] = ev[0];
+        alv[1] = ev[1];
+        alv[2] = ev[2];
```
Comment (375gnu) [inline src/audio3d/Audio3DConverts.h]: memcpy(alv, ev.data(), 3*sizeof(ALfloat)).

File: src/audio3d/Audio3DDevice.h [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+namespace Audio3D
+{
+    class Device
+    {
+        ALCdevice *_dev;
+
+    public:
+        Device(ALCdevice *dev) : _dev(dev) {}
+        Device() { _dev = alcOpenDevice(NULL); }
+
```
Comment (375gnu) [inline src/audio3d/Audio3DDevice.h]: remove copy constructor and copy operator

File: src/audio3d/Audio3DDevice.h [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        ~Device() { alcCloseDevice(_dev); }
+
+        ALCenum lastError() { return alcGetError(_dev); }
+
+        static ALCdevice *openDevice(const char *n = NULL)
+        {
+            return alcOpenDevice(n);
```
Comment (375gnu) [inline src/audio3d/Audio3DDevice.h]: not a C++ style

File: src/audio3d/Audio3DError.cpp [hunk 7]
```diff
@@ -0,0 +1,30 @@
+
+#include "Audio3DError.h" 
+
+std::string Audio3D::alErrorToString(ALenum code)
+{
+    switch (code)
+    {
+        case AL_NO_ERROR: return "AL_NO_ERROR";
```
Comment (375gnu) [inline src/audio3d/Audio3DError.cpp]:
```
case AL_...:
        return "AL_..."
```
look better
But I'd rather add a preprocessor macro 
```#define AL_ERROR_TO_STR(error) #error```

File: src/audio3d/Audio3DError.h [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    ALenum inline alError() { return alGetError(); }
+
+    ALenum inline alcError(ALCdevice *d) { return alcGetError(d); }
+
+    ALenum inline alutError() { return alutGetError(); }
+
+    void inline checkForAlError()
+    {
+        ALenum err = alError();
+        if (err != AL_NO_ERROR) {
```
Comment (375gnu) [inline src/audio3d/Audio3DError.h]: wrong style

File: src/audio3d/Audio3DManager.cpp [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    alutInitWithoutContext(argc, argv);
+}
+
+std::vector<std::string> Manager::devices()
+{
+    std::string s;
+    std::vector<std::string> l;
+    const char *p = alcGetString(0, ALC_DEVICE_SPECIFIER);
+
+    while(1)
```
Comment (375gnu) [inline src/audio3d/Audio3DManager.cpp]: In C and C++ `for(;;)` is preffered.

File: src/audio3d/Audio3DManager.cpp [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        if (*p == '\0')
+        {
+            l.push_back(s);
```
Comment (375gnu) [inline src/audio3d/Audio3DManager.cpp]: `l.emplace_back(s)`

Comment (375gnu) [reply src/audio3d/Audio3DManager.cpp]: OR `push_back` and `s.clear();` on the next line

File: src/audio3d/Audio3DSource.h [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+    {
+        ALuint _srcId;
+
+    public:
+        Source(ALuint id) : _srcId(id) {}
+        ~Source()
+        {
+            alDeleteSources(1, &_srcId);
+        }
+
```
Comment (375gnu) [inline src/audio3d/Audio3DSource.h]: delete copy constructor and copy assignment

File: src/audio3d/Audio3DSource.h [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+        ALint intProperty(ALenum param) const
+        {
+            ALint ret;
+            alGetSourcei(_srcId, param, &ret);
+            return ret;
+        }
+
+        bool looping()
+        {
+            return intProperty(AL_LOOPING) > 0 ? true : false;
```
Comment (375gnu) [inline src/audio3d/Audio3DSource.h]: just `return intProperty(AL_LOOPING) > 0;`

File: src/celengine/audio.cpp [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+#include <iostream>
+
+#include "audio.h"
+#include "audio3d/Audio3DListener.h"
+
+using namespace Audio3D;
+
+// AudioObject methods
+
+#define for_all_sources Audio3D::Source *ptr; \
```
Comment (375gnu) [inline src/celengine/audio.cpp]: facepalm.jpg

File: src/celengine/audio.cpp [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+{
+    get_first_source->looping();
+}
+
+
+
+// AudioListener methods
+
+#define for_all_ems_do \
+    for (em_it = emitMap.begin(); em_it != emitMap.end(); em_it++)
```
Comment (375gnu) [inline src/celengine/audio.cpp]: read about foreach-like form of `for` introduced in C++11

File: src/celengine/audio.h [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+
+    obs_list::const_iterator obs_it;
+    obj_list::const_iterator obj_it;
+
+    AudioObject *newObject(Selection);
+public:
+    AudioManager(int*, char **);
+
+    void setGlobalDistanceScale(float v) { gdScale = v; }
+    float globalDistanceScale() const { return gsScale; }
```
Comment (375gnu) [inline src/celengine/audio.h]: consider such style:
```
float globalDistanceScale() const { return gsScale; }
float& globalDistanceScale() { return gsScale; }
```
this should allow `foo.globalDistanceScale() = 12.34;`

File: src/audio3d/Audio3DBuffer.h (hunk 1 above)
Comment (pirogronian) [reply src/audio3d/Audio3DBuffer.h]: It started as separate project and I forgot to change it (or though it's just neglectable). So, changing now :-)

File: src/audio3d/Audio3DConverts.h (hunk 2 above)
Comment (pirogronian) [reply src/audio3d/Audio3DConverts.h]: I though about it, but was in unsure about better names. After all, this is part of small module inside its own namespace...

File: src/audio3d/Audio3DConverts.h (hunk 3 above)
Comment (pirogronian) [reply src/audio3d/Audio3DConverts.h]: I want keep it as simple (and fast) as possible. It isn't intended to be used outside of audio module.

File: src/audio3d/Audio3DBuffer.h (hunk 1 above)
Comment (375gnu) [reply src/audio3d/Audio3DBuffer.h]: I prefer `#pragma once`, that's nonstandard extension but supported by all major compilers. for ages.

File: src/audio3d/Audio3DConverts.h (hunk 2 above)
Comment (375gnu) [reply src/audio3d/Audio3DConverts.h]: actually, with `Eigen::Map<>()` and `Eigen::Vector3f::data()` you may not need this method.

Comment (pirogronian) [reply src/audio3d/Audio3DConverts.h]: Actually, it appeared all this file is unnecessary and I removed it.

File: src/audio3d/Audio3DDevice.h (hunk 5 above)
Comment (pirogronian) [reply src/audio3d/Audio3DDevice.h]: I included test code in new commit.

**Pull request closed without merging**